src/Renci.SshNet/Common/TimeSpanExtensions.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System;
+#if NET21_OR_GREATER || NET6_0_OR_GREATER
+using System.Runtime.CompilerServices;
+#endif
+
+namespace Renci.SshNet.Common
+{
+    /// <summary>
+    /// Provides extension methods for <see cref="TimeSpan"/>.
+    /// </summary>
+    internal static class TimeSpanExtensions
+    {
+        /// <summary>
+        /// Returns the specified <paramref name="timeSpan"/> as a valid timeout in milliseconds.
+        /// </summary>
+        /// <param name="timeSpan">The <see cref="TimeSpan"/> to ensure validity.</param>
+        /// <param name="callerMemberName">The name of the calling member.</param>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// Thrown when <paramref name="timeSpan"/> does not represent a value between -1 and <see cref="int.MaxValue"/>, inclusive.
+        /// </exception>
+        public static int AsTimeout(this TimeSpan timeSpan,
+#if NET21_OR_GREATER || NET6_0_OR_GREATER
+                                    [CallerArgumentExpression(nameof(timeSpan))]
+#endif
+                                    string callerMemberName = "")
+        {
+            var timeoutInMilliseconds = timeSpan.TotalMilliseconds;
+            return timeoutInMilliseconds is < -1d or > int.MaxValue
```

Comment:
What if it's -0.5?

Reply:
I'm not sure that timespan allows fractional milliseconds, I'll check.

Reply:
(int) of a double truncates the decimal. Do we want to round or is truncating ok?

Reply:
Casting to int is fine. Despite my comment, no-one should reasonably expect a timeout to conform to a fractional number of milliseconds